vnext/Microsoft.ReactNative/JsiApi.idl

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.ReactNative
+{
+  interface IJsiRuntime;
+
+  delegate void JsiDataHandler(UInt8[] data);
+
+  [webhosthidden]
+  interface IJsiBuffer
+  {
+    UInt32 Size { get; };
+    void GetData(JsiDataHandler handler);
+  };
+
+  enum JsiValueType
+  {
+    Undefined,
+    Null,
+    Boolean,
+    Number,
+    Symbol,
+    String,
+    Object,
+  };
+
+  [webhosthidden]
+  struct JsiValueData
+  {
+    JsiValueType Kind;
+    UInt64 Value;
+  };
+
+  [default_interface]
+  runtimeclass JsiPreparedJavaScript
+  {
+  };
+
+  delegate JsiValueData JsiHostFunction(IJsiRuntime runtime, JsiValueData thisValue, JsiValueData[] args);
+
+  [webhosthidden]
+  interface IJsiHostObject
+  {
+    JsiValueData GetProperty(IJsiRuntime runtime, UInt64 name);
+    void SetProperty(IJsiRuntime runtime, UInt64 name, JsiValueData value);
+    IVectorView<UInt64> GetPropertyNames(IJsiRuntime runtime);
+  };
+
+  [webhosthidden]
+  interface IJsiRuntime
+  {
+    JsiValueData EvaluateJavaScript(IJsiBuffer buffer, String sourceUrl);
+    JsiPreparedJavaScript PrepareJavaScript(IJsiBuffer buffer, String sourceUrl);
+    JsiValueData EvaluatePreparedJavaScript(JsiPreparedJavaScript js);
+    UInt64 Global { get; };
+    String Description { get; };
+    Boolean IsInspectable { get; };
+
+    UInt64 CloneSymbol(UInt64 symbol);
+    UInt64 CloneString(UInt64 str);
+    UInt64 CloneObject(UInt64 obj);
+    UInt64 ClonePropertyNameId(UInt64 propertyNameId);
+
+    UInt64 CreatePropertyNameIdFromAscii(UInt8[] ascii);
+    UInt64 CreatePropertyNameIdFromUtf8(UInt8[] utf8);
+    UInt64 CreatePropertyNameIdFromString(UInt64 str);
+    void PropertyNameIdToUtf8(UInt64 propertyNameId, JsiDataHandler utf8Handler);
+    Boolean ComparePropertyNameIds(UInt64 left, UInt64 right);
+
+    void SymbolToUtf8(UInt64 symbol, JsiDataHandler utf8Handler);
+
+    UInt64 CreateStringFromAscii(UInt8[] ascii);
+    UInt64 CreateStringFromUtf8(UInt8[] utf8);
+    void StringToUtf8(UInt64 str, JsiDataHandler utf8Handler);
+
+    JsiValueData CreateValueFromJsonUtf8(UInt8[] json);
+
+    UInt64 CreateObject();
```

Comment:
Did you consider about, instead of having JsiValueType, it can improve the type restriction by making JsiValueData to multiple structs, like JsiObjectData, JsiFunctionData, JsiArrayData, JsiArrayBufferData. All of these just have a UInt64 field. Values are not just objects, they could also be boolean (how can I know is a value true or false) or floating point numbers. Instead of passing in and out UInt64 values, you could consider about something like this:

```
JsiArrayData CreateArray(size);
JsiValueData GetValueAtIndex(JsiArrayData arr, index);
```

```
// instead of
Boolean IsArray(UInt64 obj);
// what about
Boolean TryGetArray(JsiObjectData obj, JsiArrayData* arr);
```

I think this makes the API much clear, and it still satisfy the "don't create memory if unnecessary" rule.

And I think you may have missed functions to test whether a UInt64 pointer is an object, or a string. I don't see how can I put a number to, for example, a field of an object.

Reply:
Great suggestion! I have implemented it in the iteration 6. The types like JsiObjectData or JsiArrayData encapsulate the related pointer types. We use the JsiValueData with the JisVlaueKind to encapsulate any JavaScript value such as Boolean, Number, Object, etc. The JsiValueData needs to be like a variant type. We typically do not need to encapsulate Number or Boolea types on their own.

---
In reply to: [451861640](https://github.com/microsoft/react-native-windows/pull/5461#discussion_r451861640) [](ancestors = 451861640)